amplify/backend/function/sendNotification/lib/index.ts

```diff
@@ -0,0 +1,238 @@
+// Unfortunatey we cannot destructure this library
+// https://github.com/firebase/firebase-admin-node/issues/593
+import * as admin from 'firebase-admin';
+import { Handler } from 'aws-lambda';
+
+import { graphqlRequest, notNull } from '../../utils';
+import { getParams } from '../../getParams';
+
+import {
+  Params,
+  TriggerEvent,
+  GetColonyName_SnQuery,
+  GetColonyName_SnQueryVariables,
+  GetColonyName_SnDocument,
+  GetUserDetails_SnDocument,
+  GetUserDetails_SnQueryVariables,
+  GetUserDetails_SnQuery,
+  NotificationType,
+  CreateNotification_SnMutation,
+  CreateNotification_SnMutationVariables,
+  CreateNotification_SnDocument,
+} from './types';
+
+import { notificationBuilder } from './message';
+import { sendEmail } from './email';
+
+let apiKey = 'da2-fakeApiId123456';
+let graphqlURL = 'http://localhost:20002/graphql';
+let firebaseAdminConfig;
+let mailJetApiKey;
+let mailJetApiSecret;
+
+const setEnvVariables = async () => {
+  const ENV = process.env.ENV;
+  if (ENV === 'qa') {
+    [apiKey, graphqlURL, firebaseAdminConfig, mailJetApiKey, mailJetApiSecret] =
+      await getParams([
+        'appsyncApiKey',
+        'graphqlUrl',
+        'firebaseAdminConfig',
+        'mailJetApiKey',
+        'mailJetApiSecret',
+      ]);
+  }
+};
+
+const removeInterpolationMarkers = (message: string) => {
+  return message.replace(/[\[\]]/g, '');
+};
+
+export const handler: Handler<TriggerEvent> = async (event: TriggerEvent) => {
+  try {
+    await setEnvVariables();
+  } catch (e) {
+    throw new Error(`Unable to set environment variables. Reason: ${e}`);
+  }
+
+  let params: Params = {
+    ...(event.arguments?.input || {}),
+    graphqlURL,
+    apiKey,
+    mailJetApiKey,
+    mailJetApiSecret,
+  };
+
+  const { userId, colonyId, customNotificationTitle } = params;
+
+  const serviceAccount = JSON.parse(firebaseAdminConfig);
+
+  const app = admin.initializeApp({
+    credential: admin.credential.cert(serviceAccount),
+  });
+
+  const messaging = app.messaging();
+
+  const colonyQuery = await graphqlRequest<
+    GetColonyName_SnQuery,
+    GetColonyName_SnQueryVariables
+  >(GetColonyName_SnDocument, { id: colonyId || '' }, graphqlURL, apiKey);
+
+  if (colonyQuery.errors || !colonyQuery.data) {
+    const [error] = colonyQuery.errors;
+    throw new Error(error || 'Could not fetch colony data from DynamoDB');
+  }
+
+  const { name, metadata } = colonyQuery?.data?.getColony || {};
+
+  const title =
```

Comment:
I know this was not spec'd out properly (or at all), but I'm making a case, that since notifications are app breaking, they should not have fallbacks of last resort.

So if a proper title cannot be sourced (either custom or default), the system should just stop and not send anything else out.

This is true for titles, body, names, etc, both in users and colonies

Reply:
At the moment this is true for the body, however for the title I am not convinced. 

Does a colony have to have a display name, or is this optional? Does it have to have a name or again is this optional? If we allow them to be optional (as the schema suggests) I don't think we should break notifications if they are left blank.

The same goes for the username, do we allow the username to be blank? If so, I think it is reasonable to fall back onto the user address in order to send the notification.

To me the only real breakage is if a body cannot be constructed and if all the fallbacks for title etc. fail